src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,380 @@
+        malloc((~expression).size() * sizeof(value_type))));
+    reset_pointer_vector((~expression).size());
+  } else if (not owning_) {
+    ASSERT((~expression).size() == size(), "Must copy into same size, not "
+                                               << (~expression).size()
+                                               << " into " << size());
+  }
+  ~*this = expression;
+  return *this;
+}
+/// \endcond
+
+//  declaration for copy constructor
+template <typename T, typename VectorType>
+VectorImpl<T, VectorType>& VectorImpl<T, VectorType>::operator=(
+    const VectorImpl<T, VectorType>& rhs) noexcept {
+  if (this != &rhs) {
+    if (owning_) {
+      if (size() != rhs.size()) {
+        owned_data_.reset(rhs.size() > 0 ? static_cast<value_type*>(malloc(
+                                               rhs.size() * sizeof(value_type)))
+                                         : nullptr);
+      }
+      reset_pointer_vector(rhs.size());
+    } else {
+      ASSERT(rhs.size() == size(), "Must copy into same size, not "
+                                       << rhs.size() << " into " << size());
+    }
+    std::memcpy(data(), rhs.data(), size() * sizeof(value_type));
+  }
+  return *this;
+}
+
+// declaration for move constructor
+template <typename T, typename VectorType>
+VectorImpl<T, VectorType>& VectorImpl<T, VectorType>::operator=(
+    VectorImpl<T, VectorType>&& rhs) noexcept {
+  if (this != &rhs) {
+    if (owning_) {
+      owned_data_ = std::move(rhs.owned_data_);
+      ~*this = ~rhs; /* PointerVector is trivially copyable */
+      owning_ = rhs.owning_;
+    } else {
+      ASSERT(rhs.size() == size(), "Must copy into same size, not "
+                                       << rhs.size() << " into " << size());
+      std::memcpy(data(), rhs.data(), size() * sizeof(value_type));
+    }
+    rhs.owning_ = true;
+    rhs.reset();
+  }
+  return *this;
+}
+/// \endcond
+
+// create from a single stored value, which becomes the sole entry.
+template <typename T, typename VectorType>
+VectorImpl<T, VectorType>& VectorImpl<T, VectorType>::operator=(
+    const T& rhs) noexcept {
+  ~*this = rhs;
+  return *this;
+}
+
+template <typename T, typename VectorType>
+void VectorImpl<T, VectorType>::pup(PUP::er& p) noexcept {  // NOLINT
+  auto my_size = size();
+  p | my_size;
+  if (my_size > 0) {
+    if (p.isUnpacking()) {
+      owning_ = true;
+      owned_data_.reset(my_size > 0 ? static_cast<value_type*>(
+                                          malloc(my_size * sizeof(value_type)))
+                                    : nullptr);
+      reset_pointer_vector(my_size);
+    }
+    PUParray(p, data(), size());
+  }
+}
+
+/// Output operator for VectorImpl
+template <typename T, typename VectorType>
+std::ostream& operator<<(std::ostream& os,
+                         const VectorImpl<T, VectorType>& d) noexcept {
+  sequence_print_helper(os, d.begin(), d.end());
+  return os;
+}
```

Comment:
This change actually causes ambiguous overload with blaze::DenseVector equivalence check. Do you have any objection to just removing our definitions of equivalence altogether and letting blaze handle it?

Reply:
Hmm, I think since we don't considering non-owning in the comparison using Blaze would be fine. It would be useful to note that as a comment in the Doxygen for VectorImpl, I think.